### PR TITLE
 UCP/TAG: Enable multiple tag offload ifaces

### PR DESCRIPTION
**Pull request by brminich**
- Activate offload iface when the first offload unexpected msg arrives 
- The most common case is supposed to be when tag offload messages arrive to only one iface. Consider it to be  a fast path
- Maintain a hash of (tag, iface), if several offload ifaces are activated. When some receive buffer is posted, need to do a hash lookup for the particular interface to make a post on

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5419/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@yosefe, can you please check the approach? if it is ok, I'd proceed with the tests

File: src/ucp/tag/offload.h [hunk 1]
```diff
@@ -85,4 +88,53 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
     }
 }
 
+/* For homogeneous clusters it is expected that tag offload messages will arrive
+ * from the single interface. Otherwise need to add a particular iface to tags
+ * hash, where key is a tag masked with tag_sender mask. */
+static UCS_F_ALWAYS_INLINE void
+ucp_tag_offload_unexp(ucp_worker_iface_t *iface, ucp_tag_t tag)
```
Comment (yosefe) [inline src/ucp/tag/offload.h]: looks like we have 3 extra checks for every unexp eager message, need to avoid it in case the interface is already active

File: src/ucp/tag/offload.h [hunk 2]
```diff
@@ -85,4 +88,53 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
...
+            ucs_assert(worker->tm.offload.zcopy_thresh == SIZE_MAX);
+
+            worker->tm.offload.thresh       = context->config.ext.tm_thresh;
+            worker->tm.offload.zcopy_thresh = context->config.ext.tm_max_bcopy;
+
+            ucs_debug("Enable TM offload: thresh %zu, zcopy_thresh %zu",
+                    worker->tm.offload.thresh, worker->tm.offload.zcopy_thresh);
+        }
+
+        ucp_worker_iface_activate(iface, 0);
```
Comment (yosefe) [inline src/ucp/tag/offload.h]: need to somehow unite this with the generic interface activation mechanism, when iface_progress is called explicitly when getting interrupt, and then counting active messages on the interface. i guess need to add option to modify unexp callback function after interface is already created, same as for active messages

File: src/ucp/tag/offload.c [hunk 3]
```diff
@@ -78,7 +92,8 @@ void ucp_tag_offload_completed(uct_tag_context_t *self, uct_tag_t stag,
                                UCP_RECV_DESC_FLAG_LAST);
         ucs_mpool_put_inline(req->recv.rdesc);
     } else {
-        iface = ucp_tag_offload_iface(worker);
+        iface = ucp_tag_offload_iface(worker, stag);
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: shouldn't we post on all, in case of NULL?
also i would split the hash optimization to another PR

File: src/ucp/core/ucp_worker.c [hunk 4]
```diff
@@ -666,11 +643,6 @@ static void ucp_worker_close_ifaces(ucp_worker_h worker)
             }
         }
 
-        if (ucp_worker_is_tl_tag_offload(worker, rsc_index)) {
```
Comment (alinask) [inline src/ucp/core/ucp_worker.c]: if ucp_worker_is_tl_tag_offload() isn't called from anywhere anymore, can remove this function.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5516/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5530/ for details (Mellanox internal link).

File: src/ucp/tag/offload.c (hunk 3 above)
Comment (yosefe) [inline src/ucp/tag/offload.c]: looks like merge conflict, ucp_request_recv_buffer_dereg(req) should be used here

File: src/ucp/tag/offload.c [hunk 5]
```diff
@@ -34,7 +47,8 @@ ucp_tag_offload_release_buf(ucp_request_t *req, ucp_worker_t *worker)
     if (req->recv.rdesc != NULL) {
         ucs_mpool_put_inline(req->recv.rdesc);
     } else {
-        iface = ucp_tag_offload_iface(worker);
+        iface = ucp_tag_offload_iface(worker, req->recv.tag.tag);
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: ucp_request_recv_buffer_dereg(req) should be used here

File: src/ucp/tag/offload.c [hunk 6]
```diff
@@ -181,17 +196,21 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_offload_unexp_rndv,
...
+    ucp_tag_offload_unexp(iface, stag);
 
     return UCS_OK;
 }
 
 void ucp_tag_offload_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
 {
 
-    ucp_worker_iface_t *ucp_iface = ucp_tag_offload_iface(worker);
+    ucp_worker_iface_t *ucp_iface = ucp_tag_offload_iface(worker,
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: s/ucp_iface/wiface/

File: src/ucp/tag/offload.h [hunk 7]
```diff
@@ -87,4 +90,55 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
...
+    ucp_context_t *context = worker->context;
+    ucp_tag_t tag_key;
+    khiter_t hash_it;
+    int ret;
+
+    ++iface->proxy_recv_count;
+
+    if (ucs_unlikely(!(iface->flags & UCP_WORKER_IFACE_FLAG_ON_OFFLOAD_QUEUE))) {
+
+        if (ucs_unlikely(!context->config.ext.tm_offload)) {
```
Comment (yosefe) [inline src/ucp/tag/offload.h]: the internal code block could be placed in a non-inline function to reduce register spill

File: src/ucp/tag/offload.h [hunk 8]
```diff
@@ -87,4 +90,55 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
...
+            ucs_assert(worker->tm.offload.zcopy_thresh == SIZE_MAX);
+
+            worker->tm.offload.thresh       = context->config.ext.tm_thresh;
+            worker->tm.offload.zcopy_thresh = context->config.ext.tm_max_bcopy;
+
+            ucs_debug("Enable TM offload: thresh %zu, zcopy_thresh %zu",
+                    worker->tm.offload.thresh, worker->tm.offload.zcopy_thresh);
+        }
+
+        ucs_queue_push(&worker->tm.offload.ifaces, &iface->queue);
```
Comment (yosefe) [inline src/ucp/tag/offload.h]: is there any point of keeping a queue of offload ifaces now? maybe just a single pointer in case there is single iface, and NULL if there are 0 or more than one?

File: src/ucp/core/ucp_worker.h [hunk 9]
```diff
@@ -138,8 +149,9 @@ typedef struct ucp_worker_iface {
     unsigned                      activate_count;/* How times this iface has been activated */
     int                           on_arm_list;   /* Is the interface on arm_list */
     int                           check_events_id;/* Callback id for check_events */
-    int                           proxy_am_count;/* Counts active messages on proxy handler */
-} ucp_worker_iface_t;
+    unsigned                      proxy_recv_count;/* Counts active messages on proxy handler */
+    uint8_t                       flags;         /* Interface flags */
```
[the remark this answers is not in the document]
Comment (brminich) [reply src/ucp/core/ucp_worker.h]: I was going to do that in a separate PT, because it is unrelated. Should I do it here?

File: src/ucp/tag/offload.c [hunk 10]
```diff
@@ -13,17 +13,30 @@
...
-    return ucs_queue_head_elem_non_empty(&worker->tm.offload.ifaces,
-                                         ucp_worker_iface_t, queue);
+    if (ucs_likely(worker->tm.offload.num_ifaces == 1)) {
+        iface = ucs_queue_head_elem_non_empty(&worker->tm.offload.ifaces,
+                                              ucp_worker_iface_t, queue);
+    } else {
+        key_tag = worker->context->config.tag_sender_mask & tag;
+        hash_it = kh_get(ucp_tag_offload_hash, &worker->tm.offload.tag_hash,
+                         key_tag);
+        iface   = (hash_it == kh_end(&worker->ep_hash)) ? NULL :
```
[the remark this answers is not in the document]
Comment (brminich) [reply src/ucp/tag/offload.c]: bug, will fix

File: src/ucp/tag/offload.h (hunk 8 above)
Comment (brminich) [reply src/ucp/tag/offload.h]: Is there any scenario when we can get back to single interface from several ones? Seems like no..

File: src/ucp/core/ucp_worker.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/core/ucp_worker.h]: i think yes, because this pr adds flags field so old "flags" should also be moved to use it

File: src/ucp/tag/offload.h (hunk 8 above)
Comment (yosefe) [reply src/ucp/tag/offload.h]: no, we never deactivate..

Comment (brminich) [reply src/ucp/tag/offload.h]: So will make a union of {hash/iface pointer} at this point

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5536/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5538/ for details (Mellanox internal link).

File: src/ucp/tag/tag_match.h [hunk 11]
```diff
@@ -10,12 +10,16 @@
...
 #include <ucp/core/ucp_types.h>
 #include <ucs/datastruct/queue_types.h>
+#include <ucs/datastruct/khash.h>
 #include <ucs/sys/compiler_def.h>
 
 
 #define UCP_TAG_MASK_FULL     0xffffffffffffffffUL  /* All 1-s */
 
 
+KHASH_MAP_INIT_INT64(ucp_tag_offload_hash, ucp_worker_iface_t *);
```
Comment (yosefe) [inline src/ucp/tag/tag_match.h]: need to declare a map using ucp_tag_t rather than int64_t, or assert that the types are the same (which is not 100% true..)

File: src/ucp/tag/eager_rcv.c [hunk 12]
```diff
@@ -212,11 +212,11 @@ ucs_status_t ucp_tag_offload_unexp_eager(void *arg, void *data, size_t length,
                                          uint64_t imm)
 {
     /* Align data with AM protocol. We should add tag before the data. */
-    ucp_worker_t *worker UCS_V_UNUSED = arg;
-    uint16_t flags                    = UCP_RECV_DESC_FLAG_EAGER |
-                                        UCP_RECV_DESC_FLAG_FIRST |
-                                        UCP_RECV_DESC_FLAG_LAST  |
-                                        UCP_RECV_DESC_FLAG_OFFLOAD;
+    ucp_worker_iface_t *iface = arg;
```
Comment (yosefe) [inline src/ucp/tag/eager_rcv.c]: s/iface/wiface/

File: src/ucp/tag/offload.h [hunk 13]
```diff
@@ -66,6 +67,8 @@ void ucp_tag_offload_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
 
 int ucp_tag_offload_post(ucp_request_t *req, ucp_request_queue_t *req_queue);
 
+int ucp_tag_offload_iface_activate(ucp_worker_iface_t *iface);
```
Comment (yosefe) [inline src/ucp/tag/offload.h]: pls document the return value of this function
s/iface/wiface/

File: src/ucp/tag/offload.h [hunk 14]
```diff
@@ -90,4 +93,36 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
     }
 }
 
+/* For homogeneous clusters it is expected that tag offload messages will arrive
+ * from the single interface. Otherwise need to add a particular iface to tags
+ * hash, where key is a tag masked with tag_sender mask. */
+static UCS_F_ALWAYS_INLINE void
+ucp_tag_offload_unexp(ucp_worker_iface_t *iface, ucp_tag_t tag)
```
Comment (yosefe) [inline src/ucp/tag/offload.h]: s/iface/wiface/

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5558/ for details (Mellanox internal link).

**Comment by brminich** [comment]
__builtin_types_compatible_p failed on r-vmb-sles11-sp0-x86-6 due to relatively old compiler (c99 was not enabled). Thus ULLONG_MAX was not found and khash defined khint64_t as unsigned long long.
With –std=c99, it compiles ok. 

So, removed __builtin_types_compatible_p check and  declared hash with ucp_tag_t, (but left khash hash functions, because they look appropriate)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5561/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@yosefe, comments applied. Please check and I'll squash after that

**Comment by yosefe** [comment]
👍 pls squash

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5579/ for details (Mellanox internal link).

File: src/ucp/core/ucp_worker.h [hunk 15]
```diff
@@ -44,6 +44,17 @@ enum {
 };
 
 
+/**
+ * UCP iface flags
+ */
+enum {
+    UCP_WORKER_IFACE_FLAG_OFFLOAD_ACTIVATED = UCS_BIT(0), /**< UCP iface receives tag
+                                                               offload messages */
+    UCP_WORKER_IFACE_FLAG_ON_ARM_LIST       = UCS_BIT(1)  /**< UCP iface is on the arm
```
Comment (shamisp) [inline src/ucp/core/ucp_worker.h]: do we need to clarify what is arm list ?

File: src/ucp/tag/offload.h [hunk 16]
```diff
@@ -66,6 +67,10 @@ void ucp_tag_offload_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
 
 int ucp_tag_offload_post(ucp_request_t *req, ucp_request_queue_t *req_queue);
 
+/* Returns 0 if tag offloading is disabled in the configuration,
+ * returns 1 otherwise */
```
Comment (shamisp) [inline src/ucp/tag/offload.h]: if you already document the function, why not to use doxygen style. 
Also, why not bool ?

File: src/ucp/core/ucp_worker.h (hunk 15 above)
Comment (yosefe) [reply src/ucp/core/ucp_worker.h]: yeah, maybe it's confusing with certain CPU architecture name..

File: src/ucp/tag/offload.h (hunk 16 above)
Comment (yosefe) [reply src/ucp/tag/offload.h]: +1, bool it's C++ only AFAIK

Comment (shamisp) [reply src/ucp/tag/offload.h]: C99 has within #include <stdbool.h>

Comment (brminich) [reply src/ucp/tag/offload.h]: Just was consistent, we use int when need bool in other places

**Comment by shamisp** [comment]
In ugni code we already make the assumption (c99) about bool
https://github.com/openucx/ucx/blob/51be945ed14d6a7fcbb855401a9a8e2fb518c17b/src/uct/ugni/base/ugni_types.h#L13

File: src/ucp/tag/offload.h (hunk 16 above)
Comment (shamisp) [reply src/ucp/tag/offload.h]: n ugni code we already make the assumption (c99) about bool
https://github.com/openucx/ucx/blob/51be945ed14d6a7fcbb855401a9a8e2fb518c17b/src/uct/ugni/base/ugni_types.h#L13

Comment (brminich) [reply src/ucp/tag/offload.h]: in UCP/UCT we do not use that, so I'd not change it to bool here

Comment (shamisp) [reply src/ucp/tag/offload.h]: One of the two places would have to change.

**Comment by yosefe** [comment]
bot:bgate:retest

File: src/ucp/tag/offload.h (hunk 16 above)
Comment (brminich) [reply src/ucp/tag/offload.h]: I'll update the comment. Or you meant something else?

Comment (shamisp) [reply src/ucp/tag/offload.h]: Please update comment. As for C99/bool - we would have to fix here or in ugni code.

Comment (brminich) [reply src/ucp/tag/offload.h]: I'd stay with int. Then ugni needs to be modified in a separate PR as it is not related to this feature. @yosefe , what is your opinion?

Comment (yosefe) [reply src/ucp/tag/offload.h]: int

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5592/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5594/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5595/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@shamisp, @yosefe, comments addressed. Plz, re-check

File: src/ucp/tag/offload.h (hunk 16 above)
Comment (shamisp) [reply src/ucp/tag/offload.h]: please open an bug for @MattBBaker

Comment (brminich) [reply src/ucp/tag/offload.h]: @shamisp, done - #2108
